x-pack/plugins/reporting/server/routes/generate_from_savedobject.ts

```diff
@@ -0,0 +1,157 @@
+    throw new Error(`The Export handler is expected to return a result with job info! ${result}`);
+  }
+
+  return result;
+};
+
+/*
+ * This function registers API Endpoints for queuing Reporting jobs. The API inputs are:
+ * - "immediate" flag: whether to execute the job up front and make immediate download available
+ * - saved object type and ID
+ * - time range and time zone
+ * - application state:
+ *     - filters
+ *     - query bar
+ *     - local (transient) changes the user made to the saved object
+ */
+export function registerGenerateCsvFromSavedObject(
+  server: KbnServer,
+  handleRoute: HandlerFunction,
+  handleRouteError: HandlerErrorFunction
+) {
+  const getRouteConfig = getRouteConfigFactoryReportingPre(server);
+  const routeOptions = {
+    ...getRouteConfig(() => CSV_FROM_SAVEDOBJECT_JOB_TYPE),
+    validate: {
+      params: Joi.object({
+        savedObjectType: Joi.string().required(),
+        savedObjectId: Joi.string().required(),
+      }).required(),
+      payload: Joi.object({
+        state: Joi.object().default({}),
+        timerange: Joi.object({
+          timezone: Joi.string().default('UTC'),
+          min: Joi.date().required(),
+          max: Joi.date().required(),
+        }).optional(),
+      }),
+    },
+  };
+  const getDocumentPayload = getDocumentPayloadFactory(server);
+
+  // csv: immediate download
+  server.route({
+    path: `${BASE_GENERATE}/immediate/csv/saved-object/{savedObjectType}:{savedObjectId}`,
+    method: 'POST',
+    options: routeOptions,
+    handler: async (request: Request, h: ResponseToolkit) => {
+      /*
+       * 1. Queue a job with getJobFromRouteHandler
+       *   - `isImmediate: true` gets us the complete result data in payload.objects
+       * 2. Copy the completed data stashed in the job as output.content
+       *   - Makes the content available for download
+       * 3. Return a response with CSV content
+       */
+      const queuedJob: QueuedJobPayload = await getJobFromRouteHandler(
+        handleRoute,
+        handleRouteError,
+        request,
+        h,
+        {
+          isImmediate: true,
+        }
+      );
+
+      // FIXME this is REALLY ugly
+      const jobSource: JobDoc = get(queuedJob, 'source.job');
+      const output: JobDocOutput = getDocumentPayload({
+        _source: {
+          ...jobSource,
+          status: 'completed',
+          output: {
+            content: jobSource.payload.objects,
+            content_type: 'text/csv',
+          },
+        },
+      });
+
+      const response: ResponseObject = h
+        .response(output.content)
+        .type(output.contentType)
+        .code(output.statusCode);
+
+      if (output.headers) {
+        Object.keys(output.headers).forEach(key => {
+          response.header(key, output.headers[key]);
+        });
+      }
+
+      const { isBoom } = response as KibanaResponse;
+      if (isBoom == null) {
+        response.header('accept-ranges', 'none');
+      }
```

Comment:
When returning the CSV result, this reads the entire CSV data into memory at once and sends it to the browser; it doesn't stream CSV content back.

I think this is a good solution for Phase 1 of this API. Streaming downloads could cause problems with some web browsers. 

In future plans, we can have a flag in the API request to control whether to stream or not. We can also have an Advanced Setting for CSV stream or one-shot download.

Reply:
I think we're in agreement on this to allow it to slide for now, and see if we can get streaming in V2 of this API